### PR TITLE
Separate validation from deployment

### PR DESCRIPTION
**Pull request by mproch**

### CODE REVIEW

File: engine/flink/management/periodic/src/main/scala/pl/touk/nussknacker/engine/management/periodic/PeriodicDeploymentManager.scala [hunk 1]
```diff
@@ -79,17 +79,28 @@ class PeriodicDeploymentManager private[periodic](val delegate: DeploymentManage
...
+
   override def deploy(processVersion: ProcessVersion,
                       deploymentData: DeploymentData,
                       canonicalProcess: CanonicalProcess,
                       savepointPath: Option[String]): Future[Option[ExternalDeploymentId]] = {
+    withExtractedSchedule(canonicalProcess) { scheduleProperty =>
+      logger.info(s"About to (re)schedule ${processVersion.processName} in version ${processVersion.versionId}")
+      // PeriodicProcessStateDefinitionManager do not allow to redeploy (so doesn't GUI),
+      // but NK API does, so we need to handle this situation.
+      service.schedule(scheduleProperty, processVersion, canonicalProcess, cancelIfJobPresent(processVersion, deploymentData.user))
```
[the remark this answers is not in the document]
Comment (mproch) [reply engine/flink/management/periodic/src/main/scala/pl/touk/nussknacker/engine/management/periodic/PeriodicDeploymentManager.scala:99]: :+1: changed a bit

File: ui/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentService.scala [hunk 2]
```diff
@@ -77,16 +79,21 @@ class DeploymentService(processRepository: FetchingProcessRepository[Future],
   private def deployAndSaveProcess(process: BaseProcessDetails[CanonicalProcess],
                                    savepointPath: Option[String],
                                    deploymentComment: Option[DeploymentComment],
-                                   performDeploy: (ProcessingType, ProcessVersion, DeploymentData, CanonicalProcess, Option[String]) => Future[_])(implicit user: LoggedUser): Future[ProcessActionEntityData] = {
-    for {
+                                   deploymentManager: DeploymentManager)(implicit user: LoggedUser): Future[Future[ProcessActionEntityData]] = {
```
[the remark this answers is not in the document]
Comment (mproch) [reply ui/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentService.scala:83]: yes, added comment for that...

File: ui/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/DeploymentManager.scala [hunk 3]
```diff
@@ -11,24 +11,33 @@ import scala.concurrent.Future
 
 trait DeploymentManager extends AutoCloseable {
 
-  //TODO: savepointPath is very flink specific, how can we handle that differently?
+  /**
+    * This method is invoked separately before deploy, to be able to give user quick feedback, as deploy (e.g. on Flink) may take long time
+    */
+  def validate(processVersion: ProcessVersion, deploymentData: DeploymentData, canonicalProcess: CanonicalProcess): Future[Unit]
```
Comment (dswiecki) [inline ui/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/DeploymentManager.scala:17]: why do we return future here ? I mean we rather return cats Validated when when we do synchronous validation

Comment (mproch) [reply ui/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/DeploymentManager.scala:17]: it's not synchronous (at least not necessarily)

**Review by gadomsky — APPROVED** [review]
LGTM

**Pull request merged**